### PR TITLE
Use clearDataForOrigin instead of forceUpdateServiceWorkers

### LINKED ISSUE #592: Migrate to Storage.clearDataForOrigin for clearing storage & SW
Do we want to [clear storage](https://chromedevtools.github.io/debugger-protocol-viewer/tot/Storage/) before we gather against a URL for the first time? Feels destructive (esp for the extension!), but we already clear SWs and force them to reinstall, and I'm figuring we could get skewed data if stuff already exists in LS, IDB, Caches etc.

Perhaps a config option to clear data before run? wdyt?

### PR DESCRIPTION
**Pull request by wardpeet**
Fixes #592

### CODE REVIEW

File: lighthouse-core/lib/storage-types.js [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+ * limitations under the License.
+ */
+'use strict';
+
+/**
+ * Enum of possible storage types.
+ * @enum {string}
+ * @see https://chromedevtools.github.io/debugger-protocol-viewer/tot/Storage/#type-StorageType
+ */
+const StorageTypes = {
```
Comment (brendankenny) [inline lighthouse-core/lib/storage-types.js]: is this file necessary? If we're only ever going to use `'all'`, it would probably be easier to use it directly

File: lighthouse-core/gather/gather-runner.js [hunk 2]
```diff
@@ -89,8 +89,8 @@ class GatherRunner {
       .then(_ => {
         return driver.cleanAndDisableBrowserCaches();
       }).then(_ => {
-        // Force SWs to update on load.
-        return driver.forceUpdateServiceWorkers();
+        // Clears storage for origin.
+        return driver.clearDataForOrigin(options.url);
```
Comment (brendankenny) [inline lighthouse-core/gather/gather-runner.js]: does this need to be the origin, not the full URL? or will the command deal with it for you

File: lighthouse-core/gather/gatherers/service-worker.js [hunk 3]
```diff
@@ -24,6 +24,12 @@ class ServiceWorker extends Gatherer {
     return driver
       .getServiceWorkerVersions()
       .then(data => {
+        if (hasMultipleOrigins(data.versions)) {
```
Comment (brendankenny) [inline lighthouse-core/gather/gatherers/service-worker.js]: did you mean to include this change? (seems like it's #601)

**Comment by brendankenny** [comment]
weirdly the smoketests are failing on this, but only in node v5+, not node v4 with `--harmony`. Just restarted and they failed again. 

```
> lighthouse-cli/scripts/run-smoke-tests.sh
> method <= browser ERR:error Storage.clearDataForOrigin +0ms 
> Runtime error encountered: { code: -32601,
  message: '\'Storage.clearDataForOrigin\' wasn\'t found' }
> undefined
> Fail! Lighthouse run didn't create a result file
```

**Comment by wardpeet** [comment]
somehow closure script is failing locally on the master branch as well..

File: lighthouse-core/lib/storage-types.js (hunk 1 above)
Comment (wardpeet) [inline lighthouse-core/lib/storage-types.js]: sure thing :)

File: lighthouse-core/gather/gatherers/service-worker.js (hunk 3 above)
Comment (wardpeet) [inline lighthouse-core/gather/gatherers/service-worker.js]: ugh correct, doing too much at once :D

**Comment by brendankenny** [comment]
looks like the protocol issue was an old cache.

But deleting that, now I get an issue with linting the extension, only on the v4 bot :) Weirdly node can find `eslint` and `eslint-config-google` one directory up, but can't find `eslint-config-xo`, even though that's required by `eslint-config-google`. Maybe this is an eslint/v4 issue?

**Comment by wardpeet** [comment]
will have a look into it tomorrow and test it locally

File: lighthouse-core/gather/gather-runner.js (hunk 2 above)
Comment (wardpeet) [inline lighthouse-core/gather/gather-runner.js]: updated inside driver

File: lighthouse-core/gather/drivers/driver.js [hunk 4]
```diff
@@ -420,11 +421,21 @@ class Driver {
...
   }
 
-  forceUpdateServiceWorkers() {
-    // COMPAT: This command will trigger this registrationId error in Chrome 50 (51 undetermined):
-    //   "{"code":-32602,"message":"Missing or invalid 'registrationId' parameter"}"
-    return this.sendCommand('ServiceWorker.setForceUpdateOnPageLoad', {
-      forceUpdateOnPageLoad: true
+  clearDataForOrigin(url) {
+    const parsedUrl = parseURL(url);
+    const origin = `${parsedUrl.protocol}//${parsedUrl.hostname}`;
```
Comment (brendankenny) [inline lighthouse-core/gather/drivers/driver.js]: need the port number too, if there is one. Something like [this](https://github.com/GoogleChrome/lighthouse/blob/a379d8f4f1633962163e55d5144ffe80bdbd4982/lighthouse-core/audits/service-worker.js#L30)

File: lighthouse-core/gather/drivers/driver.js [hunk 5]
```diff
@@ -420,11 +421,21 @@ class Driver {
...
+
+    if (!parsedUrl.protocol) {
```
Comment (brendankenny) [inline lighthouse-core/gather/drivers/driver.js]: this should be caught earlier, if we're going to check the url. Maybe in runner.js, with [the https check](https://github.com/GoogleChrome/lighthouse/blob/a379d8f4f1633962163e55d5144ffe80bdbd4982/lighthouse-core/runner.js#L49)?

File: lighthouse-core/gather/drivers/driver.js [hunk 6]
```diff
@@ -420,11 +421,21 @@ class Driver {
...
+      throw new Error(`Couldn't find protocol of ${url}`);
+    }
+
+    if (!parsedUrl.hostname) {
+      throw new Error(`Couldn't find hostname of ${url}`);
```
Comment (brendankenny) [inline lighthouse-core/gather/drivers/driver.js]: same with this check

File: lighthouse-core/test/lib/drivers/driver-test.js [hunk 7]
```diff
@@ -32,6 +32,8 @@ driverStub.sendCommand = function(command, params) {
       return Promise.resolve({
         nodeId: params.selector === 'invalid' ? 0 : 231
       });
+    case 'Storage.clearDataForOrigin':
+      return Promise.resolve({origin: 'http://aliexpress.com', storageTypes: 'all'});
```
Comment (brendankenny) [inline lighthouse-core/test/lib/drivers/driver-test.js]: the actual command doesn't appear to have a return object, so I don't think this should resolve on anything, just `return Promise.resolve();`

File: lighthouse-core/test/lib/drivers/driver-test.js [hunk 8]
```diff
@@ -80,4 +82,23 @@ describe('Browser Driver', () => {
     assert.notEqual(opts.url, req1.url, 'opts.url changed after the redirects');
     assert.equal(opts.url, req3.url, 'opts.url matches the last redirect');
   });
+
+  it('will clear origin of a page', () => {
```
Comment (brendankenny) [inline lighthouse-core/test/lib/drivers/driver-test.js]: is this something we can really test? Pretty much all of it is up to the browser to get right, so I'm not sure if there's any value in testing it

File: lighthouse-core/test/lib/drivers/driver-test.js (hunk 7 above)
Comment (wardpeet) [inline lighthouse-core/test/lib/drivers/driver-test.js]: this was to make sure to see if my origin parsing was correct XD

File: lighthouse-core/test/lib/drivers/driver-test.js (hunk 8 above)
Comment (wardpeet) [inline lighthouse-core/test/lib/drivers/driver-test.js]: See [comment above](https://github.com/GoogleChrome/lighthouse/pull/632#discussion_r78193037)
I added this to see if my origin got calculated right

Comment (wardpeet) [inline lighthouse-core/test/lib/drivers/driver-test.js]: Should I remove the test?

Comment (paulirish) [inline lighthouse-core/test/lib/drivers/driver-test.js]: This test doesn't enter into the driver code at all, just stays in this file. 

----

One possible way to "test" this is to instantiate a real driver, but then override sendCommand. We could than call `driver.clearDataForOrigin` and we'd at least be able to determine that we parsed an origin correctly

but that's really all we'd be testing. IMO not worth it, so we should just nuke the test, yeah.
